notebooks/notebook_template_review.py

```diff
@@ -268,73 +279,112 @@ def init():
         NotebookRule.costs = ''
     
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, cells: list) -> None:
         pass
+    
+    def get_cell(self) -> list:
+        """
+            Get the next notebook cell.
+
+            Returns:
+
+            cell: content of the next cell
+        """
+        while self.empty_cell():
+            NotebookRule.cell_index += 1
```

Comment:
You are modifying the class definition, instead of the actual instance of the class. 

This needs to be `self.cell_index`.

Better yet, remove `cell_index` altogether.

Reply:
I use the static version of 'cell_index' instead of the instance version to maintain tracking where I am in the notebook. If I used the instance version, I believe the location would be lost when the next Rule is executed.

Reply:
I don't think you need `get_cell`, just call pass in cells to `validate` and loop over them.

Reply:
I see two approaches. One we can retain location knowledge in the abstract base class (current). The other is to retain location knowledge outside of the class. This would require additional tracking code between the rules. While I can see the benefit in that, your example of using a generator (prior PR) would be very messy.

The third choice would make finding the cells location independent. Then the derived case could ask the abstract class to go find the cell. BUT, you would loose cell order dependence. Perhaps one could make a dictionary of call order after the rules have been applied, and then do a pass over the dictionary checking order.

The third approach (I used to be a compiler writer early in my career) would be my preference. But, that's a lot of engineering vs the benefit of coding the most elegant solution.

Reply:
done